### PR TITLE
Introduce action status scoped custom code

### PR DESCRIPTION
**Pull request by herdt-michael**
Allow providing a custom code with an action status feedback to give more fine grained device specific details.

Provide optional fields for DDI and DMF API to provide custom feedback with an action feedback.

Signed-off-by: Michael Herdt <Michael.Herdt@bosch.io>

### CODE REVIEW

File: hawkbit-dmf/hawkbit-dmf-amqp/src/main/java/org/eclipse/hawkbit/amqp/AmqpMessageHandlerService.java [hunk 1]
```diff
@@ -355,6 +355,11 @@ private void updateActionStatus(final Message message) {
         final ActionStatusCreate actionStatus = entityFactory.actionStatus().create(action.getId()).status(status)
                 .messages(messages);
 
+        actionUpdateStatus.getCode().ifPresent(code -> {
+            actionStatus.code(code);
+            actionStatus.message("Device reported code: " + code);
```
Comment (stefbehl) [inline hawkbit-dmf/hawkbit-dmf-amqp/src/main/java/org/eclipse/hawkbit/amqp/AmqpMessageHandlerService.java:360]: Naming the field "code" is fine for me. How about using "status code" within the message to make it more tangible: "Device reported status code: ..."

File: hawkbit-rest/hawkbit-ddi-api/src/test/java/org/eclipse/hawkbit/ddi/json/model/DdiStatusTest.java [hunk 2]
```diff
@@ -55,19 +55,47 @@ public void shouldSerializeAndDeserializeObject() throws IOException {
                 ddiStatus.getResult().getProgress().getCnt());
         assertThat(deserializedDdiStatus.getResult().getProgress().getOf()).isEqualTo(
                 ddiStatus.getResult().getProgress().getOf());
+        assertThat(deserializedDdiStatus.getDetails()).isEqualTo(ddiStatus.getDetails());
+    }
+    
+    private static Stream<Arguments> ddiStatusPossibilities(){
+        DdiProgress ddiProgress = new DdiProgress(30, 100);
```
Comment (stefbehl) [inline hawkbit-rest/hawkbit-ddi-api/src/test/java/org/eclipse/hawkbit/ddi/json/model/DdiStatusTest.java:62]: declar local variables "final"

File: hawkbit-rest/hawkbit-ddi-api/src/test/java/org/eclipse/hawkbit/ddi/json/model/DdiStatusTest.java [hunk 3]
```diff
@@ -55,19 +55,47 @@ public void shouldSerializeAndDeserializeObject() throws IOException {
...
     @Test
     @Description("Verify the correct deserialization of a model with a additional unknown property")
     public void shouldDeserializeObjectWithUnknownProperty() throws IOException {
         // Setup
         String serializedDdiStatus = "{\"execution\":\"proceeding\",\"result\":{\"finished\":\"none\","
-                + "\"progress\":{\"cnt\":30,\"of\":100}},\"details\":[],\"unknownProperty\":\"test\"}";
+              + "\"progress\":{\"cnt\":30,\"of\":100}},\"details\":[],\"unknownProperty\":\"test\"}";
+
+        // Test
+        DdiStatus ddiStatus = mapper.readValue(serializedDdiStatus, DdiStatus.class);
```
Comment (stefbehl) [inline hawkbit-rest/hawkbit-ddi-api/src/test/java/org/eclipse/hawkbit/ddi/json/model/DdiStatusTest.java:78]: declare local variables "final"

**Review by stefbehl — CHANGES_REQUESTED** [review]
Looks good. Thanks @herdt-michael !
Could you please also enhance org.eclipse.hawkbit.rest.mgmt.documentation.TargetResourceDocumentationTest to make sure that the new status code field is also documented as part of the Management REST API?

File: hawkbit-dmf/hawkbit-dmf-amqp/src/main/java/org/eclipse/hawkbit/amqp/AmqpMessageHandlerService.java (hunk 1 above)
Comment (herdt-michael) [reply hawkbit-dmf/hawkbit-dmf-amqp/src/main/java/org/eclipse/hawkbit/amqp/AmqpMessageHandlerService.java:360]: adapted

File: hawkbit-rest/hawkbit-ddi-api/src/test/java/org/eclipse/hawkbit/ddi/json/model/DdiStatusTest.java (hunk 3 above)
Comment (herdt-michael) [reply hawkbit-rest/hawkbit-ddi-api/src/test/java/org/eclipse/hawkbit/ddi/json/model/DdiStatusTest.java:78]: declared

**Comment by herdt-michael** [comment]
@stefbehl thank you for the review! I have adapted the org.eclipse.hawkbit.rest.mgmt.documentation.TargetResourceDocumentationTest and for sure extended the MGMT definition of the Action Status to add the code within the payload.

**Review by stefbehl — APPROVED** [review]
Looks good. Thanks @herdt-michael for contributing this new feature!

**Pull request merged**